src/databases/indexeddb.ts

```diff
@@ -26,21 +27,38 @@ export default class IndexedDBDatabase implements BaseDatabase {
         db.createObjectStore("customEvents", { autoIncrement : true });
         db.createObjectStore("timingEvents", { autoIncrement : true });
       },
+    }).catch(() => {
+      console.warn(
+        "Could not open IndexedDB database to store telemetry events. This session events won't be recorded."
+      );
+      return new MemoryDatabase();
     });
   }
 
   public async addCustomEvent(eventType: string, customEvent: object) {
+    const db = await this.dbPromise;
+    if (db instanceof MemoryDatabase) {
```

Comment:
This is a bit unconventional: we're now resolving the `dbPromise` Promise with either an `IndexedDB` connection object or with a `MemoryDatabase` object depending on whether `IndexedDb` could be opened.

It's not beautiful, but thanks to the typing system it's safe and no error prone, so I think it's fine 🙃

Reply:
Is there some way to DRY up all this repetition behind a helper method of some sort?

Reply:
I don't know if we can DRY this, since it's just a method call and a conditional, and since we can't reuse the call inside the conditional (at least it does not occur to me how could we do it), there's not much to extract easily.

Only way would be to do some dynamic programming (by creating the class methods dynamically), but that would mess with TypeScript (and prefer the explicitness of having normal methods in the class).